ext/opentelemetry-ext-jaeger/src/opentelemetry/ext/jaeger/__init__.py

```diff
@@ -0,0 +1,369 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jaeger Span Exporter for OpenTelemetry."""
+
+import logging
+import socket
+import typing
+
+from thrift.protocol import TBinaryProtocol, TCompactProtocol
+from thrift.transport import THttpClient, TTransport
+
+import opentelemetry.trace as trace_api
+from opentelemetry.ext.jaeger.gen.agent import Agent as agent
+from opentelemetry.ext.jaeger.gen.jaeger import Collector as jaeger
+from opentelemetry.sdk.trace.export import Span, SpanExporter, SpanExportResult
+
+DEFAULT_AGENT_HOST_NAME = "localhost"
+DEFAULT_AGENT_PORT = 6831
+DEFAULT_COLLECTOR_ENDPOINT = "/api/traces?format=jaeger.thrift"
+
+UDP_PACKET_MAX_LENGTH = 65000
+
+logger = logging.getLogger(__name__)
+
+
+class JaegerSpanExporter(SpanExporter):
+    """Jaeger span exporter for OpenTelemetry.
+
+    Args:
+        service_name: Service that logged an annotation in a trace.Classifier
+            when query for spans.
+        agent_host_name: (Optional) The host name of the Jaeger-Agent.
+        agent_port: (Optional) The port of the Jaeger-Agent.
+        collector_host_name: (Optional) The host name of the Jaeger-Collector
+            HTTP Thrift.
+        collector_port: (Optional) The port of the Jaeger-Collector HTTP
+            Thrift.
+        collector_endpoint: (Optional) The endpoint of the Jaeger-Collector
+            HTTP Thrift.
+        username: (Optional) The user name of the Basic Auth if authentication
+            is required.
+        password: (Optional) The password of the Basic Auth if authentication
+            is required.
+    """
+
+    def __init__(
+        self,
+        service_name: str = "my_service",
+        agent_host_name: str = DEFAULT_AGENT_HOST_NAME,
+        agent_port: int = DEFAULT_AGENT_PORT,
+        collector_host_name: str = None,
+        collector_port: int = None,
+        collector_endpoint: str = DEFAULT_COLLECTOR_ENDPOINT,
+        username: str = None,
+        password: str = None,
+    ):
+        self.service_name = service_name
+        self.agent_host_name = agent_host_name
+        self.agent_port = agent_port
+        self._agent_client = None
+        self.collector_host_name = collector_host_name
+        self.collector_port = collector_port
+        self.collector_endpoint = collector_endpoint
+        self.username = username
+        self.password = password
+        self._collector = None
+
+    @property
+    def agent_client(self):
+        if self._agent_client is None:
```

Comment:
What do other people think about threadsafety here? Is it safe to assume exporters aren't threadsafe unless they specifically say otherwise?

Reply:
According to the specs: `Export() will never be called concurrently for the same exporter instance. Export() can be called again only after the current call returns.`. So it should not be a problem. A different issue is that the span processors we have doesn't fully comply with that part of the specs.

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-tracing.md#exportbatch